src/crush/CrushLocation.cc

```diff
@@ -103,8 +103,15 @@ int CrushLocation::init_on_startup()
     return update_from_hook();
   }
 
-  // start with a sane default
-  char hostname[HOST_NAME_MAX + 1];
+  /*
+   * The gethostname returns null-terminated hostname.
+   * Array needs to accommodate HOST_NAME_MAX + \0.
+   * In case of truncation done by gethostname it is unspecified
+   * whether the returned buffer includes a terminating null byte.
+   * Allocate additional element which always will be set to 0
+   * to prevent exiting boundary of the array during iteration.
+   */
+  char hostname[HOST_NAME_MAX + 2] = {0};
```

Comment:
i think the correct fix is just 
```diff
diff --git a/src/crush/CrushLocation.cc b/src/crush/CrushLocation.cc
index 0e5a0855cd..9b02b00733 100644
--- a/src/crush/CrushLocation.cc
+++ b/src/crush/CrushLocation.cc
@@ -105,7 +105,7 @@ int CrushLocation::init_on_startup()

   // start with a sane default
   char hostname[HOST_NAME_MAX + 1];
-  int r = gethostname(hostname, sizeof(hostname)-1);
+  int r = gethostname(hostname, sizeof(hostname));
   if (r < 0)
     strcpy(hostname, "unknown_host");
   // use short hostname
```

as we already take the terminating null byte into consideration.

Reply:
That was my first approach but then I started to think about situation described in gethostname manual which says that if truncation happen it is not ensure that returned array will contain null termination. I'm not sure if it possible to get from gethostname array with 64 chars without null as last element, if yes then iteration which happen below can exit array boundaries if we allow gethostname to replace all elements in hostname array.

Reply:
please note, in GETHOSTNAME(2)

>      POSIX.1  guarantees  that  "Host names  (not  including  the
>      terminating null byte) are  limited to HOST_NAME_MAX bytes".
>      On Linux, HOST_NAME_MAX is defined  with the value 64, which
>      has been the limit since  Linux 1.0 (earlier kernels imposed
>      a limit of 8 bytes).
> 

and

>    C library/kernel differences
>      The GNU C  library does not employ  the gethostname() system
>      call;  instead, it  implements  gethostname()  as a  library
>      function that calls uname(2) and copies up to len bytes from
>      the returned nodename field into name.  Having performed the
>      copy, the function then checks if the length of the nodename
>      was greater  than or equal  to len, and  if it is,  then the
>      function returns -1 with errno  set to ENAMETOOLONG; in this
>      case,  a  terminating  null  byte is  not  included  in  the
>      returned name.

so we can ensure that `hostname` is large enough for holding the longest hostname allowed by POSIX.1. 

> I'm not sure if it possible to get from gethostname array with 64 chars without null as last element

with my patch the terminating null byte is always included even when it comes to 64-char hostname. actually, you can try my suggested change with the testcase you put in https://github.com/ceph/ceph/pull/32947#issue-368324218 .

Reply:
Yes my thinking was the same, but then I got doubts after reading this parts:

> If the null-terminated hostname is too large to fit, then the name is truncated, and no
       error is returned (but see NOTES below).  POSIX.1 says that if such
       truncation occurs, then it is unspecified whether the returned buffer
       includes a terminating null byte.

Reply:
I mean this part make me think that there is possibility that gethostname during this truncation may return non null-terminated string, and if we will pass to the gethostname whole hostname array we may end up with hostname array without null symbol, that is why a I wanted to be prepared for that kind of situation. But maybe I misunderstood this part, what do you think?

Reply:
@michaelndn i think you missed 

> POSIX.1  guarantees  that  "Host names  (not  including  the
>  terminating null byte) are  limited to HOST_NAME_MAX bytes".
>  On Linux, HOST_NAME_MAX is defined  with the value 64, which
>  has been the limit since  Linux 1.0 (earlier kernels imposed
>  a limit of 8 bytes).

that's why i believe that truncation never happens as (sorry for repeating):

we can ensure that hostname is large enough for holding the longest hostname allowed by POSIX.1.

see also
```c++
#ifndef HOST_NAME_MAX
#ifdef MAXHOSTNAMELEN
#define HOST_NAME_MAX MAXHOSTNAMELEN
#else
#define HOST_NAME_MAX 255
#endif
#endif
```